bigtable/bttest/grpc_interceptor.go

```diff
@@ -0,0 +1,294 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+This file provides interceptors to hook into grpc streaming calls
+and inject latency and errors into cbtemulator for testing
+*/
+package bttest
+
+import (
+	"fmt"
+	"log"
+	"math/rand"
+	"strconv"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	validStreamMethodSuffixes = []string{"ReadRows", "MutateRows"}
+)
+
+/*
+	Creates interceptors inject latency or errors into cbtemulator
+*/
+type EmulatorInterceptorBuilder struct {
+	LatencyTargets       latencyTargets
+	GrpcErrorCodeTargets grpcErrorCodeTargets
+}
+
+func (esib *EmulatorInterceptorBuilder) BuildStreamInterceptor() grpc.ServerOption {
+	return grpc.StreamInterceptor(esib.interceptorFunc())
+}
+
+func (esib *EmulatorInterceptorBuilder) interceptorFunc() grpc.StreamServerInterceptor {
+	fmt.Println("Building Stream Server Interceptor:")
+	if len(esib.LatencyTargets) > 0 {
+		fmt.Printf(" - Latency Targets: %s\n", esib.LatencyTargets.String())
+	}
+	if len(esib.GrpcErrorCodeTargets) > 0 {
+		fmt.Printf(" - Error Targets: %s\n", esib.GrpcErrorCodeTargets.String())
+	}
+
+	return func(srv interface{},
+		ss grpc.ServerStream,
+		info *grpc.StreamServerInfo,
+		handler grpc.StreamHandler) error {
+
+		startTime := time.Now()
+
+		// Latency injection
+		pVal := rand.Int31n(100)
+		for _, lt := range esib.LatencyTargets {
+			if strings.HasSuffix(info.FullMethod, lt.methodSuffix) && pVal >= lt.percentile {
```

Comment:
I think the list needs to be sorted for this work. Otherwise something like `ReadRows:p50:10ms`, `ReadRows:p99:100ms` will never experience p99 latency

Reply:
I don't think we need to sort latency. The code iterates through all LatencyTargets regardless of whether a match has happened.  With `ReadRows:p50:10ms`, `ReadRows:p99:100ms`, we'll sleep for ~10ms then sleep for ~90ms.

Errors definitely should be sorted, since the code will return when it finds a matching target.  Added new function/test to correctly handle this case